clang/include/clang/AST/ExprCXX.h

```diff
@@ -158,9 +158,9 @@ class CXXOperatorCallExpr final : public CallExpr {
                : getOperatorLoc();
   }
 
-  SourceLocation getBeginLoc() const { return Range.getBegin(); }
-  SourceLocation getEndLoc() const { return Range.getEnd(); }
-  SourceRange getSourceRange() const { return Range; }
+  SourceLocation getBeginLoc() const { return BeginLoc; }
+  SourceLocation getEndLoc() const { return getSourceRangeImpl().getEnd(); }
```

Comment:
I think we could simplify getEndLoc() to be a trimmed-down version of getSourceRangeImpl

* for OO_Call/ OO_Subscript => `getRParenLoc()`
* for  1 argument = > `getArg(0)->getEndLoc()`
* for OO_PlusPlus/OO_MinusMinus => `getOperatorLoc()`
* for  2 argument: `getArg(1)->getEndLoc()`


I am not sure about the current handling of `OO_Arrow` in `getSourceRangeImpl`

Reply:
We could do this, and apply the same approach to `getBeginLocImpl` as well. However, I’m not sure it’s a good idea to split `getSourceRangeImpl` across two places -- it makes the logic harder to follow and reason about.

I actually tried this approach in an earlier version of the patch, but didn’t observe any performance benefit. Personally, I prefer the current version, which keeps all the logic centralized in getSourceRangeImpl for better maintainability and clarity.